File: classes/RoutesViewController.m
//
//  SchedulesViewController.m
//  marguerite
//
//  Created by Kevin Conley on 9/18/13.
//  Copyright (c) 2013 Cardinal Devs. All rights reserved.
//

#import "RoutesViewController.h"
#import "ScheduleViewController.h"
#import "MRoute.h"

@interface RoutesViewController ()

@end

@implementation RoutesViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    allRoutes = [MRoute getAllRoutes];
    
    // Sort the routes alphabetically by name
    allRoutes = [allRoutes sortedArrayUsingComparator:^NSComparisonResult(id a, id b) {
        MRoute *firstRoute = (MRoute *) a;
        MRoute *secondRoute = (MRoute *) b;
        
        return [firstRoute.routeLongName caseInsensitiveCompare:secondRoute.routeLongName];
    }];
    
    [self.tableView reloadData];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    // Return the number of sections.
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    // Return the number of rows in the section.
    return [allRoutes count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *cellIdentifier = @"AllRoutesCell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellIdentifier forIndexPath:indexPath];
    
    MRoute *route = [allRoutes objectAtIndex:indexPath.row];
    
    cell.textLabel.text = route.routeLongName;
    cell.textLabel.textColor = route.routeTextColor;
    cell.contentView.backgroundColor = route.routeColor;
    
    return cell;
}

#pragma mark - Table view delegate

- (void)tableView:(UITableView *)tableView willDisplayCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath {
    cell.backgroundColor = cell.contentView.backgroundColor;
}

#pragma mark - Segue

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    ScheduleViewController *scheduleViewController = segue.destinationViewController;
    NSIndexPath *indexPath = [self.tableView indexPathForCell:(UITableViewCell*)sender];
    scheduleViewController.schedule = [[MSchedule alloc] initWithRoute:[allRoutes objectAtIndex:indexPath.row]];
}

@end

File: Marguerite/RoutesViewController.swift
//
//  RoutesViewController.swift
//  Marguerite
//
//  Created by Kevin Conley on 3/12/15.
//  Copyright (c) 2015 Kevin Conley. All rights reserved.
//

import UIKit

class RoutesViewController: UIViewController, UITableViewDataSource, UITableViewDelegate {

    // MARK: - Model

    var routes = [Route]()

    // MARK: - Outlets

    @IBOutlet weak var tableView: UITableView!

    // MARK: - View Controller Lifecycle

    override func viewDidLoad() {
        super.viewDidLoad()

        // Get all routes, sorting them alphabetically (case-insensitive)
        routes = Route.getAllRoutes().sorted {
            $0.displayName.lowercaseString < $1.displayName.lowercaseString }
    }

    // MARK: - Storyboard Constants

    private struct Storyboard {
        static let routeTableViewCellIdentifier = "RouteTableCell"
    }

    // MARK: - UITableViewDataSource

    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return routes.count
    }

    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier(Storyboard.routeTableViewCellIdentifier, forIndexPath: indexPath) as UITableViewCell
        let route = routes[indexPath.row]
        cell.textLabel?.text = route.displayName
        cell.textLabel?.textColor = route.routeTextColor
        cell.backgroundColor = route.routeColor

        return cell
    }

    // MARK: - Segues

    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        if let wvc = segue.destinationViewController as? WebViewController {
            if let indexPath = tableView.indexPathForSelectedRow() {
                let route = routes[indexPath.row]
                wvc.urlToLoad = getURLForRoute(route)
                wvc.hideToolbar = true
            }
        }
    }

    // MARK: - Route Maps

    /**
    Some of the routes have bogus URLs, so handle those cases with this function.

    :param: route The route.

    :returns: The URL for the map of the given route.
    */
    private func getURLForRoute(route: Route) -> NSURL? {
        switch route.routeShortName {
        case "RMH":
            // This route doesn't actually seem to exist on the Marguerite website...
            return nil
        case "W":
            return NSURL(string: "http://transportation.stanford.edu/marguerite/w/map.pdf")
        case "MC-Hol":
            return NSURL(string: "http://transportation.stanford.edu/marguerite/mch/map.pdf")
        default:
            return NSURL(string: "\(route.routeUrl.absoluteString!)/map.pdf")
        }
    }
}
